janusgraph-lucene/src/main/java/org/janusgraph/diskstorage/lucene/LuceneIndex.java

```diff
@@ -221,50 +217,261 @@ public void mutate(Map<String, Map<String, IndexMutation>> mutations, KeyInforma
+            if (cardinality == Cardinality.LIST) {
+                for (IndexableField indexableField : list) {
+                    doc.add(indexableField);
+                }
+            } else {
+                // cardinality == Cardinality.SET
+                if (!encounteredSetValues.containsKey(fieldName)) encounteredSetValues.put(fieldName, new HashSet<>());
+                final Set<FieldPair> set = encounteredSetValues.get(fieldName);
+                for (IndexableField indexableField : list) {
+                    // only add the field to the doc if it hasn't already been added
+                    FieldPair fieldPair = fieldPair(indexableField);
+                    if (!set.contains(fieldPair)) {
+                        doc.add(indexableField);
+                        set.add(fieldPair);
                     }
                 }
+            }
+        }
 
-                addToDocument(storeName, documentId, doc, mutation.getAdditions(), geoFields, information);
+        // at this point the document should only hold stored (persistent) IndexableFields
+        // now add (non-persistent) fields that are used for lookups in the document (except for DOCID)
+        // the copy is necessary to avoid concurrent modifications
+        final List<IndexableField> fields = doc.getFields();
+        final List<IndexableField> copy = new ArrayList<>(fields.size());
+        copy.addAll(fields);
+        for (IndexableField field : copy) {
+            if (!field.name().equals(DOCID)) addNonPersistantField(doc, storename, field, informations);
+        }
+
+        writer.updateDocument(new Term(DOCID, docid), doc);
+    }
 
-                //write the old document to the index with the modifications
-                writer.updateDocument(new Term(DOCID, documentId), doc);
+    // we should be able to restore all fields (except DOCID) from an old document, which means they must have a stored counterpart
+    private IndexableField storedField(final String fieldName, final Object value, boolean isCardinalitySingle, final Function<String, Mapping> mappingFactory) {
+        if (value instanceof Number) {
+            if (AttributeUtil.isWholeNumber((Number) value)) {
+                return new StoredField(fieldName, ((Number) value).longValue());
+            } else { //double or float
+                return new StoredField(fieldName, ((Number) value).doubleValue());
             }
-            writer.commit();
-        } finally {
-            IOUtils.closeQuietly(reader);
+        } else if (AttributeUtil.isString(value)) {
+            String str = (String) value;
+            // we only need a mapping when we have a String value
+            Mapping mapping = mappingFactory.apply(fieldName);
+            switch (mapping) {
+                case DEFAULT:
+                case TEXT:
+                    return new TextField(fieldName, str.toLowerCase(), Field.Store.YES);
```

Comment:
why toLowerCase?

Reply:
because, in the Janusgraph specification, STRING search is case sensitive 
but TEXT search should be case insensitive and a way to achieve this is to lowercase all tokens

> When a string property is indexed as text, only full-text search predicates are supported in graph queries by the indexing backend. Full-text search is case-insensitive
quotation taken from  http://docs.janusgraph.org/latest/index-parameters.html

Actually, the fact that TEXT search should be case insensitive (and lowering the case) is In my opinion a bit silly because : 
1) it harms performence a bit 
2) it transforms the strings
3) it makes no sense for non-english language (what is lower case for japanese ?)
4) normalization of the text should be taken care by the analyzer (and by the user before putting it in a janus graph)
But as it is in the janusgraph specification so I implemented it like this (and I had to, to make tests pass in a previous pr : this comes from the pr implementing custom analyzer support)